### DISCUSSION TITLE
Reusing workflow in private repo (from same repo)

### DISCUSSION THREAD

**Discussion by iautom8things** [community/community#25719] [Actions]
<p>I know that Reusable Workflows are hamstrung by the fact that they <a href="https://github.community/t/support-reusable-workflows-in-private-repositories/212833">cannot</a> be <a href="https://github.community/t/reusable-workflows-in-private-organization-repositories/215009">reused</a> across <a href="https://github.community/t/is-there-anything-i-need-to-do-to-enable-reusable-workflows-beta/209765/5">private</a> <a href="https://github.community/t/github-actions-reusing-workflows/207163">repositories</a> within an organization.</p>
<p>But as it’s referenced a few times in the above links and the error message, workflows <em>should</em> be reusable in a private repository if the caller/called workflow is in the same repo:</p>
<blockquote>
<ul>
<li>Reusable workflows stored within a private repository can only be used by workflows within the same repository.</li>
</ul>
</blockquote>
<p>From <a href="https://docs.github.com/en/actions/learn-github-actions/reusing-workflows" rel="noopener nofollow ugc">the docs</a>.</p>
<p>I cannot get this to work.  Could someone please show me an example of this working, or point out what I’m doing wrong.  I’ve scrubbed+simplified the example below so know that I’m not actually using the literal value <code>org/repo</code> in my workflows.</p>
<pre><code class="lang-yaml"># caller.yaml
---
name: Build Images

on:
  pull_request:
    branches: [main]
    types: [closed]
jobs:
  build:
    name: Bake Image
    if: github.event_name == 'pull_request' &amp;&amp; github.event.pull_request.merged_by != ''
    uses: org/repo/.github/workflows/build_images.yaml@main
    secrets:
      SECRET: ${{ secrets.SECRET }}
</code></pre>
<pre><code class="lang-yaml"># build_images.yaml
---
name: Triggered Build Images

on:
  workflow_call:
    secrets:
      SECRET: { required: true }

jobs:
  build:
    name: Bake Image
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      # ...
</code></pre>
<p>This yields the following error:</p>
<pre><code class="lang-auto">Invalid workflow file : .github/workflows/caller.yaml#L12
error parsing called workflow "org/repo/.github/workflows/build_images.yaml@main": workflow was not found. See https://docs.github.com/actions/learn-github-actions/reusing-workflows#access-to-reusable-workflows for more information.
</code></pre>
<p>These two files sit right next to each other in the same repo:</p>
<pre><code class="lang-bash">ls -ahl .github/workflows
# build_images.yaml caller.yaml
</code></pre>

**Comment by iautom8things** [comment]
<p>I’m convinced this just doesn’t work.</p>
<p>I’ve created a public repo that has a trivial example for reusable workflows:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/iautom8things/test_reusable_workflows">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/iautom8things/test_reusable_workflows" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/522040/181092808-2160706b-72ec-4c74-9b0b-2409e475f1c1.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/4/5/454ee309d025ca31ba5082217806f97b827badbf_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/4/5/454ee309d025ca31ba5082217806f97b827badbf_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/4/5/454ee309d025ca31ba5082217806f97b827badbf.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/4/5/454ee309d025ca31ba5082217806f97b827badbf_2_10x10.png"></div>

<h3><a href="https://github.com/iautom8things/test_reusable_workflows" target="_blank" rel="noopener nofollow ugc">GitHub - iautom8things/test_reusable_workflows</a></h3>

  <p>Contribute to iautom8things/test_reusable_workflows development by creating an account on GitHub.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>If I create a private repo and copy these two files and simply change the repo name to the newly created private repo, it fails:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/a/8/a895e8ad5db8713fdfed6ec2fbc3e70709babf0d.jpeg" data-download-href="/uploads/short-url/o3noo5ieuEXU4VtYODShGHbBN0x.jpeg?dl=1" title="Screen Shot 2021-12-22 at 10.08.56 AM" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/522040/181092809-e9edb6dc-98da-4ce3-b39e-bba524719831.jpeg" alt="Screen Shot 2021-12-22 at 10.08.56 AM" data-base62-sha1="o3noo5ieuEXU4VtYODShGHbBN0x" width="690" height="300" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/a/8/a895e8ad5db8713fdfed6ec2fbc3e70709babf0d_2_690x300.jpeg, https://aws1.discourse-cdn.com/github/optimized/3X/a/8/a895e8ad5db8713fdfed6ec2fbc3e70709babf0d_2_1035x450.jpeg 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/a/8/a895e8ad5db8713fdfed6ec2fbc3e70709babf0d_2_1380x600.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/a/8/a895e8ad5db8713fdfed6ec2fbc3e70709babf0d_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">Screen Shot 2021-12-22 at 10.08.56 AM</span><span class="informations">1920×837 48.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div><br>
<div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/8/2/82db7f3bba1029164c01c93eaba5e9c6a475d2ee.png" data-download-href="/uploads/short-url/iFCla2OJHTghL11ZjsrU4SXpO46.png?dl=1" title="Screen Shot 2021-12-22 at 10.08.42 AM" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/522040/181092810-a335581b-849b-4fb7-8bdd-3278bc9c0b5c.png" alt="Screen Shot 2021-12-22 at 10.08.42 AM" data-base62-sha1="iFCla2OJHTghL11ZjsrU4SXpO46" width="656" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/8/2/82db7f3bba1029164c01c93eaba5e9c6a475d2ee_2_656x500.png, https://aws1.discourse-cdn.com/github/optimized/3X/8/2/82db7f3bba1029164c01c93eaba5e9c6a475d2ee_2_984x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/8/2/82db7f3bba1029164c01c93eaba5e9c6a475d2ee_2_1312x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/8/2/82db7f3bba1029164c01c93eaba5e9c6a475d2ee_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">Screen Shot 2021-12-22 at 10.08.42 AM</span><span class="informations">1546×1178 90.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>So either I’m <em>missing something</em>, which is certainly possible.  Or the documentation is wrong and resusable workflows <em>just don’t work with private repositories <strong>.</strong></em></p>

**Comment by schmidtw** [comment]
<p>I had a similar issue with a corporate internal repo.  Apparently the <code>uses</code> line is case sensitive and must match what github thinks it is to work.</p>
<p>This post tipped me off:</p><aside class="onebox stackexchange" data-onebox-src="https://stackoverflow.com/questions/70003863/reusing-github-action-workflow-steps-inside-another-job">
  <header class="source">

      <a href="https://stackoverflow.com/questions/70003863/reusing-github-action-workflow-steps-inside-another-job" target="_blank" rel="noopener nofollow ugc">stackoverflow.com</a>
  </header>

  <article class="onebox-body">
      <a href="https://stackoverflow.com/users/4136777/duncan-luk" target="_blank" rel="noopener nofollow ugc">
    <img alt="Duncan Luk" src="https://user-images.githubusercontent.com/698825/181092814-92c1eda6-f188-4d59-846c-5695efc3b4d6.png" class="thumbnail onebox-avatar" width="256" height="256">
  </a>

<h4>
  <a href="https://stackoverflow.com/questions/70003863/reusing-github-action-workflow-steps-inside-another-job" target="_blank" rel="noopener nofollow ugc">Reusing GitHub Action workflow steps inside another job</a>
</h4>

<div class="tags">
  <strong>github, continuous-integration, github-actions</strong>
</div>

<div class="date">
  asked by
  
  <a href="https://stackoverflow.com/users/4136777/duncan-luk" target="_blank" rel="noopener nofollow ugc">
    Duncan Luk
  </a>
  on <a href="https://stackoverflow.com/questions/70003863/reusing-github-action-workflow-steps-inside-another-job" target="_blank" rel="noopener nofollow ugc">11:48AM - 17 Nov 21 UTC</a>
</div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Not sure if that is what you’re running into, but wanted to share in case it was.</p>
<p>Wes</p>

**Comment by iautom8things** [comment]
<p>Thank you.  I don’t believe this is my issue since both my username and the repository names are all lowercase.</p>
<p>I’ve come to the conclusion that Github has no intentions of making these features available and usable to non-enterprise users.</p>

**Comment by ViacheslavKudinov** [comment]
<p>Hi,<br>
its not exactly how you tried it before, but they announced local usage.</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.blog/changelog/2022-01-25-github-actions-reusable-workflows-can-be-referenced-locally/">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/56436734/181092818-063a9455-db0c-4cdc-80f9-0da297b0b670.png" class="site-icon" width="32" height="32">

      <a href="https://github.blog/changelog/2022-01-25-github-actions-reusable-workflows-can-be-referenced-locally/" target="_blank" rel="noopener nofollow ugc">The GitHub Blog</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://github.blog/wp-content/uploads/2022/01/Product-Cloud-Enterprise@2x.png?fit=2400%2C1260" class="thumbnail" width="690" height="362"></div>

<h3><a href="https://github.blog/changelog/2022-01-25-github-actions-reusable-workflows-can-be-referenced-locally/" target="_blank" rel="noopener nofollow ugc">GitHub Actions: Reusable workflows can be referenced locally |</a></h3>

  <p>GitHub Actions: Reusable workflows can be referenced locally</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
<p>
Maybe if to use not org path but local one it will work for you.</p>

**Comment by ViacheslavKudinov** [comment, marked as answer]
<p>I think its only two ways how it could work<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/7/7/7723de86b14adb0f1fd8e7546c33fa503f956846.png" data-download-href="/uploads/short-url/gZXFrDSCIMacFH0TCy5y7VyMgAe.png?dl=1" title="Screenshot 2022-01-26 at 18.50.26" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/56436734/181092819-1dc0a955-3559-4a91-b5c9-96a9e45483ef.png" alt="Screenshot 2022-01-26 at 18.50.26" data-base62-sha1="gZXFrDSCIMacFH0TCy5y7VyMgAe" width="690" height="105" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/7/7/7723de86b14adb0f1fd8e7546c33fa503f956846_2_690x105.png, https://aws1.discourse-cdn.com/github/optimized/3X/7/7/7723de86b14adb0f1fd8e7546c33fa503f956846_2_1035x157.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/7/7/7723de86b14adb0f1fd8e7546c33fa503f956846_2_1380x210.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/7/7/7723de86b14adb0f1fd8e7546c33fa503f956846_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">Screenshot 2022-01-26 at 18.50.26</span><span class="informations">1496×228 38.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div><br>
if not to talk about Enterprise where it can be used from <code>internal</code> repo:s as well.

**Reply by DmitryKolesnikov-KSO** [reply]
Our organization is private. Can we reuse workflows within our organization between different repos?

**Reply by dataviruset** [reply]
This should now be possible between private repos rather than internal ones (which needs GitHub Enterprise): https://github.blog/changelog/2022-12-14-github-actions-sharing-actions-and-reusable-workflows-from-private-repositories-is-now-ga/

**Comment by ViacheslavKudinov** [comment]
Hi @kolesnikovd 
If you use one of Free, Pro or Team then reusable workflow has to be in same or public repo.
From doc:
> A reusable workflow can be used by another workflow if either of the following is true:
> 
> Both workflows are in the same repository.
> The called workflow is stored in a public repository, and your organization allows you to use public reusable workflows.

https://docs.github.com/en/actions/using-workflows/reusing-workflows#access-to-reusable-workflows
For Enterprise it can be also in Internal repo.
https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/reusing-workflows#access-to-reusable-workflows

**Reply by DmitryKolesnikov-KSO** [reply]
I confirmed we have Enterprise. I'll keep trying... not sure what I'm doing wrong yet, but I'm following directions. Thank you for your response.

**Reply by madhaviannavajula** [reply]
I have a similar issue . we use enterprise and still unable to figure out what is the problem. Were you able to solve this ?

**Reply by DmitryKolesnikov-KSO** [reply]
> I have a similar issue . we use enterprise and still unable to figure out what is the problem. Were you able to solve this ?

Yes, we were able to sort this out. We have one repo acting as a source of workflows and many repos inheriting.

Source repo uses `on: workflow_call` and inheriting repo has the `uses` keyword under "jobs" and references the workflow file

**Comment by areg-gareginyan** [comment]
This is possible now , need to update the access settings of private repo that holds the workflow to allow access across organization - https://docs.github.com/en/actions/creating-actions/sharing-actions-and-workflows-from-your-private-repository
